pkg/traffic/gateway.go

```diff
@@ -10,19 +10,38 @@ import (
 	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
 
 	"github.com/Kuadrant/multicluster-gateway-controller/pkg/_internal/slice"
+	"github.com/Kuadrant/multicluster-gateway-controller/pkg/controllers/gateway"
 )
 
 type GatewayInterface interface {
 	Interface
 	GetListenerByHost(host string) *gatewayv1beta1.Listener
 }
 
-func NewGateway(g *gatewayv1beta1.Gateway) GatewayInterface {
-	return &Gateway{Gateway: g}
+func NewGateway(g *gatewayv1beta1.Gateway) (*Gateway, error) {
+	gw := &Gateway{Gateway: g, isMultiCluster: false}
+
+	for i, address := range gw.Status.Addresses {
+		if i == 0 {
+			gw.isMultiCluster = isMultiClusterAddressType(*address.Type)
+			continue
+		}
+		if gw.isMultiCluster == isMultiClusterAddressType(*address.Type) {
+			continue
+		}
+		return nil, fmt.Errorf("gateway is invalid: inconsistent status addresses")
```

Comment:
Not really following this logic, can you explain what it's doing?

Reply:
So it goes through the list of all the addresses and ensures they are the same type. 
The first address sets the boolean flag on the wrapper; all addresses after this are compared against this flag. In case the address type and previously set flag don't agree we have an error